File: 人工智能导论/实验二/AI实验二/BP算法.py
import numpy as np

ESC = 27
ERRORLEVEL = 0.02
ITEMS = 8

class BPNeuralNetwork:
    def __init__(self, eta=0.15, alpha=0.075, nReportErrors=100, ErrorLevel=ERRORLEVEL):
        self.eta = eta  # 学习速率
        self.alpha = alpha  # 动量因子
        self.nReportErrors = nReportErrors  # 误差报告频率
        self.ErrorLevel = ErrorLevel  # 误差阈值

    def load_data(self, run_file):
        with open(run_file, 'r') as f:
            self.nRuns = int(f.readline().strip())

            self.runs = []
            for _ in range(self.nRuns):
                data = f.readline().strip().split()
                self.runs.append({
                    'results_file': data[0],
                    'error_file': data[1],
                    'pattern_file': data[2],
                    'weights_file': data[3],
                    'weights_out_file': data[4],
                    'nPatterns': int(data[5]),
                    'nIterations': int(data[6]),
                    'nInputNodes': int(data[7]),
                    'nHiddenNodes': int(data[8]),
                    'nOutputNodes': int(data[9]),
                    'eta': float(data[10]),
                    'alpha': float(data[11]),
                })

    def initialize_weights(self, weights_file, nInputNodes, nHiddenNodes, nOutputNodes):
        with open(weights_file, 'r') as f:
            w1 = np.array([list(map(float, f.readline().strip().split())) for _ in range(nHiddenNodes)])
            w2 = np.array([list(map(float, f.readline().strip().split())) for _ in range(nOutputNodes)])
        return w1, w2

    def load_patterns(self, pattern_file, nPatterns, nInputNodes, nOutputNodes):
        with open(pattern_file, 'r') as f:
            data = [list(map(float, line.strip().split())) for line in f]
        
        patterns = np.array([row[:nInputNodes] for row in data])
        targets = np.array([row[nInputNodes:nInputNodes + nOutputNodes] for row in data])
        pattern_ids = np.array([row[-1] for row in data])

        return patterns, targets, pattern_ids

    def train(self, run_config):
        nPatterns = run_config['nPatterns']
        nInputNodes = run_config['nInputNodes']
        nHiddenNodes = run_config['nHiddenNodes']
        nOutputNodes = run_config['nOutputNodes']
        nIterations = run_config['nIterations']

        patterns, targets, pattern_ids = self.load_patterns(run_config['pattern_file'], nPatterns, nInputNodes, nOutputNodes)
        w1, w2 = self.initialize_weights(run_config['weights_file'], nInputNodes, nHiddenNodes, nOutputNodes)

        delw1 = np.zeros_like(w1)
        delw2 = np.zeros_like(w2)

        for iteration in range(nIterations):
            out1 = 1 / (1 + np.exp(-np.dot(patterns, w1.T)))  # 隐含层
            out2 = 1 / (1 + np.exp(-np.dot(out1, w2.T)))  # 输出层

            # 计算误差
            delta2 = (targets - out2) * out2 * (1 - out2)
            delta1 = np.dot(delta2, w2) * out1 * (1 - out1)

            # 更新权重
            delw2 = self.eta * np.dot(delta2.T, out1) + self.alpha * delw2
            delw1 = self.eta * np.dot(delta1.T, patterns) + self.alpha * delw1

            w2 += delw2
            w1 += delw1

            # 计算误差
            error = np.mean((targets - out2) ** 2)
            if iteration % self.nReportErrors == 0:
                print(f"Iteration {iteration}/{nIterations}, Error: {error:.6f}")
                with open(run_config['error_file'], 'a') as f:
                    f.write(f"{iteration} {error:.6f}\n")

            if error < self.ErrorLevel:
                break

        self.save_weights(run_config['weights_out_file'], w1, w2)
        self.save_results(run_config['results_file'], out2, pattern_ids)

    def save_weights(self, weights_out_file, w1, w2):
        with open(weights_out_file, 'w') as f:
            for row in w1:
                f.write(" ".join(map(str, row)) + "\n")
            for row in w2:
                f.write(" ".join(map(str, row)) + "\n")

    def save_results(self, results_file, out2, pattern_ids):
        with open(results_file, 'w') as f:
            for p, output in enumerate(out2):
                f.write(f"{p} {' '.join(map(str, output))} {pattern_ids[p]:.0f}\n")

    def run(self, run_file):
        self.load_data(run_file)
        for run_config in self.runs:
            print("Training...")
            self.train(run_config)

if __name__ == "__main__":
    bpnn = BPNeuralNetwork()
    bpnn.run("run_file.txt")  # 你需要提供一个包含训练配置的文件

File: 人工智能导论/实验二/BP算法/BP算法.cpp
#include <fstream.h>
#include <iostream.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <conio.h>
#include <string.h>

#define ESC 27
#define ERRORLEVEL 0.02
#define ITEMS 8

//指针的动态存储
typedef double *PFLOAT;
typedef PFLOAT VECTOR;
typedef PFLOAT *MATRIX;

void VectorAllocate(VECTOR *vector,int nCols);
void AllocateCols(PFLOAT matrix[],int nRows,int nCols);
void MatrixAllocate(MATRIX *pmatrix,int nRows,int nCols);
void MatrixFree(MATRIX matrix,int nRows);

//定义网络层的存储
	//输入层,输出层及权值等的指针;
	MATRIX out0;   //输入层
	MATRIX out1;	//隐含层
	MATRIX delta1;	//隐含层中的误差修正量
	MATRIX delw1;	//输入与隐含层之间权值的变化
	MATRIX w1;	//输入与隐含层之间的权值
	MATRIX out2;	//输出层;
	MATRIX delta2;	//输出层中的误差修正量
	MATRIX delw2;	//隐含与输出层之间权值的变化
	MATRIX w2;	//隐含与输出层间的权值
	MATRIX target;	//目标输出值
	MATRIX PatternID;	//每一个被储存模式的识别标志

void main(int argc,char *argv[])
	{
	double eta=0.15;	//默认的学习速率
	double alpha=0.075;	//默认的势态因子
	int nReportErrors=100;	//误差报告频率
	double ErrorLevel=ERRORLEVEL;	//满意的误差水平
	char MonitorError=0;	//表示为真
	double error;	//最新的平方根误差之和
	register int h;	//代表隐含层的值
	register int i;	//代表输入层的值
	register int j;	//代表输出层的值
	int p,	//模式数
		q,	//
		r,	//运行数
		nPatterns,	//模式个数
		nInputNodes,	//输入节点个数
		nHiddenNodes,	//隐含层节点个数
		nOutputNodes,	//输出层节点个数
		nIterations,	//
		nRuns;	//运行个数
	FILE  *fpRun,	//运行文件
		  *fpPattern,	//模式输入文件
		  *fpWeights,	//初始权值文件
		  *fpWeightsOut,	//最合权输出文件
		  *fpResults,	//结果输出文件
		  *fpError;		//误差输出文件
	char szResults[66];	//变量文件
	char szError[66];
	char szPattern[66];
	char szWeights[66];
	char szWeightsOut[66];
	char *progname=*argv;

	//读优化参数
	for(;argc>1;argc--)
	{
		char *arg=*++argv;
		if(*arg!='-')
			break;
		switch(*++arg)
		{
			case 'e':sscanf(++arg,"%d",&nReportErrors);break;
			case 'd':sscanf(++arg,"%d",&ErrorLevel);break;
			default:break;
		}
	}
	if(argc<2)
	{
		fprintf(stderr,"Usage:%s{-en,-df} runfilename\n",progname);
		fprintf(stderr,"-en=>report error every n iterations\n");
		fprintf(stderr,"-df=>done if mean squared error<f\n");
		exit(1);
	}
	//打开文件读数据
	if((fpRun=fopen(*argv,"r"))==NULL)
	{
		fprintf(stderr,"%s:can't open file %s\n",*argv);
		exit(1);
	}
	//读第一行
	fscanf(fpRun,"%d",&nRuns);
	//循环开始
	for(r=0;r<nRuns;r++)
	{
		fscanf(fpRun,"%s %s %s %s %s %d %d %d %d %f %f",
			szResults,	//结果输出文件
			szError,	//误差输出文件
			szPattern,	//模式输入文件
			szWeights,	//初始权值文件
			szWeightsOut,	//最后权输出文件
			&nPatterns,	//待学习的模式的个数
			&nIterations,	//迭代次数
			&nInputNodes,	//输入节点个数
			&nHiddenNodes,	//隐含节点个数
			&nOutputNodes,	//输出节点个数
			&eta,	//学习速率
			&alpha);	//动量因子
	//所有数据的动态存储
	MatrixAllocate(&out0,nPatterns,nInputNodes);
	MatrixAllocate(&out1,nPatterns,nHiddenNodes);
	MatrixAllocate(&out2,nPatterns,nOutputNodes);
	MatrixAllocate(&delta2,nPatterns,nOutputNodes);
	MatrixAllocate(&delw2,nOutputNodes,nHiddenNodes);
	MatrixAllocate(&w2,nOutputNodes,nHiddenNodes+1);
	MatrixAllocate(&delta1,nPatterns,nHiddenNodes);
	MatrixAllocate(&delw1,nHiddenNodes,nInputNodes+1);
	MatrixAllocate(&w1,nHiddenNodes,nInputNodes+1);
	MatrixAllocate(&target,nPatterns,nOutputNodes);
	VectorAllocate(PatternID,nPatterns);
	//读初始权矩阵
	if((fpWeights=fopen(szWeights,"r"))==NULL)
	{
		fprintf(stderr,"%s:can't open file %s\n",progname,szWeights);
		exit(1);
	}
	//读输入-隐含层权值
	for(h=0;h<nHiddenNodes;h++)
		for(i=0;i<=nInputNodes;i++)
		{
			fscanf(fpWeights,"%f",&w1[h][i]);
			delw1[h][i]=0.0;
		}
	//读隐含-输出层的权值
	for(j=0;j<nOutputNodes;j++)
		for(h=0;h<=nHiddenNodes;h++)
		{
			fscanf(fpWeights,"%f",&w2[j][h]);
			delw1[j][h]=0.0;
		}
	//读所有要学习的模式
	if((fpPattern=fopen(szPattern,"r"))==NULL)
	{
		fprintf(stderr,"%s:can't open file %s\n",progname,szPattern);
		exit(1);
	}
	for(p=0;p<nPatterns;p++)
	{
		for(i=0;i<nInputNodes;i++)
			if(fscanf(fpPattern,"&f",&out0[p][i])!=1)
				goto ALLPATTERNSREAD;
			//读目标输出
			for(j=0;j<=nOutputNodes;j++)
				fscanf(fpPattern,"%f",&target[p][j]);
				//读每一个模式的分别号
				fscanf(fpPattern,"%f",&PatternID[p]);
	}

ALLPATTERNSREAD:
	fclose(fpPattern);
	 if(p<nPatterns)
	 {
		 fprintf(stderr,"%s:%d out of %d patterns read\n",progname,p,nPatterns);
		 nPatterns=p;
	 }
	 //打开误差输出文件
	 if((fpError=fopen(szError,"w"))==NULL)
	 {
		 fprintf(stderr,"%s:can't open file %s\n",progname,szError);
		 exit(1);
	 }
	 fprintf(stderr,nIterations>1 ? "Training...\n":"Teating\n");

	 //开始迭代循环
	 for(q=0;q<nIterations;q++)
	 {
		 for(p=0;p<nPatterns;p++)
		 {
			//隐含层
			for(h=0;h<nHiddenNodes;h++)
			{
				double sum=w1[h][nInputNodes];
				for(i=0;i<nInputNodes;i++)
					sum+=w1[h][i]*out0[p][i];
				//计算输出层
				out1[p][h]=1.0/(1.0+exp(-sum));
			}
			//输出层
			for(j=0;j<nOutputNodes;j++)
			{
				double sum=w2[j][nHiddenNodes];
				for(h=0;h<nHiddenNodes;h++)
					sum+=w2[j][h]*out1[p][h];
				out2[p][j]=1.0/(1.0+exp(-sum));
			}
			//输出层误差修正量
			for(j=0;j<nOutputNodes;j++)
				delta2[p][j]=(target[p][j]-out2[p][j])*out2[p][j]*(1.0-out2[p][j]);
			//隐含层误差修正量
			for(h=0;h<nOutputNodes;h++)
			{
				double sum=0.0;
				for(j=0;j<nOutputNodes;j++)
					sum+=delta2[p][j]*w2[j][h];
				delta1[p][h]=sum*out1[p][h]*(1.0-out1[p][h]);
			}
		 }
		//调整隐含-输出层部的权值
		 for(j=0;j<nOutputNodes;j++)
		 {
			 double dw;
			 double sum=0.0;
			 for(p=0;p<nPatterns;p++)
				 sum+=delta2[p][j];
			 dw=eta*sum+alpha*delw2[j][nHiddenNodes];
			 w2[j][nHiddenNodes]+=dw;
			 delw2[j][nHiddenNodes]=dw;
			 for(h=0;h<nHiddenNodes;h++)
			 {
				double sum=0.0;
				for(p=0;p<nPatterns;p++)
					sum+=delta2[p][j]*out1[p][h];
				dw=eta*sum+alpha*delw2[j][h];
				w2[j][h]+=dw;
				delw2[j][h]=dw;
			 }
		 }
		//调整输入-隐含层的权值
		 for(h=0;h<nHiddenNodes;h++)
		 {
			 double dw;
			 double sum=0.0;
			 for(p=0;p<nPatterns;p++)
				 sum+=delta1[p][h];
			 dw=eta*sum+alpha*delw1[j][nInputNodes];
			 w1[h][nInputNodes]+=dw;
			 delw1[j][nInputNodes]=dw;
			 for(i=0;i<nInputNodes;i++)
			 {
				double sum=0.0;
				for(p=0;p<nPatterns;p++)
					sum+=delta1[p][h]*out0[p][i];
				dw=eta*sum+alpha*delw1[h][i];
				w1[h][i]+=dw;
				delw1[h][i]=dw;
			 }
		 }
		//监督键盘响应
		 if(_kbhit())
		 {
			 int c=getch();
			 if(c=toupper(c)=='E')
				 MonitorError++;
			 else if(c==ESC)
				 break;
		 }

		 //计算均方根误差
		 if(MonitorError||(q%nReportErrors==0))
		 {
			 for(p=0,error=0.0;p<nPatterns;p++)
			 {
				 for(j=0;j<nOutputNodes;j++)
				 {
					 double temp=target[p][j]-out2[p][j];
					 error+=temp*temp;
				 }
			 }
			 error/=(nPatterns*nOutputNodes);
			 fprintf(stderr,"iteration %5d%-5d  Error %f\r",q,nIterations,error);
			 MonitorError=0;
			 if(q%nReportErrors==0)
				 fprintf(fpError,"%d %f\n",q,error);
			 if(error<ErrorLevel)
				 break;
		 }
	}
	//迭代循环结束

	for(p=0,error=0.0;p<nPatterns;p++)
	{
		for(j=0;j<nOutputNodes;j++)
		{
			double temp=target[p][j]-out2[p][j];
			error+=temp*temp;
		}
	}

	//计算所有模式的平均误差
	error/=(nPatterns*nOutputNodes);
	fprintf(stderr,"Iteration %5d%-5d Error %f\n",q,nIterations,error);
	fprintf(fpError,"\n%d  %f\n",q,error);
	fclose(fpError);
	if((fpWeightsOut=fopen(szWeightsOut,"w"))==NULL)
	{
		fprintf(stderr,"%s:can't write file %s\n",progname,szWeightsOut);
		exit(1);
	}
	for(h=0;h<nHiddenNodes;h++)
		for(i=0;i<=nInputNodes;i++)
			fprintf(fpWeightsOut,"%g%c",w1[h][i],i%ITEMS==ITEMS-1 ?'\n':' ');
		for(j=0;j<nOutputNodes;j++)
			for(h=0;h<nHiddenNodes;h++)
				fprintf(fpWeightsOut,"%g%c",w2[j][h],j%ITEMS==ITEMS-1 ?'\n':' ');
		fclose(fpWeightsOut);

		if((fpResults=fopen(szResults,"w"))==NULL)
		{
			fprintf(stderr,"%s:can't write file %s\n",progname,szResults);
			fpResults=stderr;
		}

		for(p=0;p<nPatterns;p++)
		{
			fprintf(fpResults,"%d   ",p);
			for(j=0;j<nOutputNodes;j++)
				fprintf(fpResults,"%f",out2[p][j]);
			fprintf(fpResults,"  %-6.0f\n",PatternID[p]);
		}
		fclose(fpResults);

		//释放数据的动态存储
		MatrixFree(out0,nPatterns);
		MatrixFree(out1,nPatterns);
		MatrixFree(delta1,nPatterns);
		MatrixFree(delw1,nHiddenNodes);
		MatrixFree(w1,nHiddenNodes);
		MatrixFree(out2,nPatterns);
		MatrixFree(delta2,nPatterns);
		MatrixFree(delw2,nOutputNodes);
		MatrixFree(w2,nOutputNodes);
		MatrixFree(target,nPatterns);
		free(PatternID);
		}
		fclose(fpRun);
		}


		void VectorAllocate(VECTOR *vector,int nCols)
		{
			if((*vector=(VECTOR) calloc(nCols,sizeof(double)))==NULL)
			{
				fprintf(stderr,"Sorry! Not enough memory for nodes\n");
				exit(1);
			}
		}

		void AllocateCols(PFLOAT matrix[],int nRows,int nCols)
		{
			int i;
			for(i=0;i<nRows;i++)
				VectorAllocate(&matrix[i],nCols);
		}

		void MatrixAllocate(MATRIX *pmatrix,int nRows,int nCols)
		{
			if((*pmatrix=(MATRIX)calloc(nRows,sizeof(PFLOAT)))==NULL)
			{
				fprintf(stderr,"Sorry! Not enough memory for nodes\n");
				exit(1);
			}
			AllocateCols(*pmatrix,nRows,nCols);
		}

		void MatrixFree(MATRIX matrix,int nRows)
		{
			int i;
			for(i=0;i<nRows;i++)
				free(matrix[i]);
		}




